### PR TITLE
Add command "oc login --token"

### PR DESCRIPTION
**Pull request by slav9n4ik**

### CODE REVIEW

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    public OpenShiftCredentials(String username, String password) {
+        this.username = username;
+        this.password = password;
+    }
+
+    public OpenShiftCredentials(String token) {
+        this.token = token;
+    }
+
+    public static OpenShiftCredentials of (String credentialsString) {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java:24]: remove space

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        String[] split = credentialsString.split(":");
+        if (split.length == 2) return new OpenShiftCredentials(split[0], split[1]);
+        if (split.length == 1) return new OpenShiftCredentials(split[0]);
+        return null;
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java:28]: do not return null from public methods

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 3]
```diff
@@ -3,30 +3,55 @@
...
+
+    }
+
+    private static OpenShiftProject getBuildWithUsernamePassword(OCExecutor oc, OSDFState state, OpenShiftProperties properties) {
+        return OpenShiftProject.builder()
+                .clusterUrl(properties.clusterUrl())
+                .project(properties.project())
+                .username(state.getOpenShiftCredentials().getUsername())
+                .password(state.getOpenShiftCredentials().getPassword())
+                .oc(oc).build();
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java:42]: put build at new line

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 4]
```diff
@@ -40,6 +65,10 @@ private void login() {
         oc.execute("oc login " + clusterUrl + " -u " + username + " -p " + password);
     }
 
+    private void loginWithToken() {
+        oc.execute("oc login --token=" + token);
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java:69]: maybe add cluster url?

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package io.microconfig.osdf.state;
+
+import lombok.*;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@EqualsAndHashCode
+public class OpenShiftCredentials {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java:9]: do not copy logic from Credentials class

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 6]
```diff
@@ -3,30 +3,55 @@
...
+    private final String token;
 
     private final OCExecutor oc;
 
     public static OpenShiftProject create(OSDFPaths paths, OCExecutor oc) {
         OSDFState state = fromFile(paths.stateSavePath());
         OpenShiftProperties properties = properties(propertyGetter(state.getEnv(), paths.configPath()));
-        return new OpenShiftProject(properties.clusterUrl(), state.getOpenShiftCredentials().getUsername(),
-                state.getOpenShiftCredentials().getPassword(), properties.project(), oc);
+        if (StringUtils.isEmpty(state.getOpenShiftCredentials().getToken())) {
```
[the remark this answers is not in the document]
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: +

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 7]
```diff
@@ -3,30 +3,55 @@
...
+            return getBuildWithUsernamePassword(oc, state, properties);
+        } else {
+            return getBuildWithToken(oc, state, properties);
+        }
+
+    }
+
+    private static OpenShiftProject getBuildWithUsernamePassword(OCExecutor oc, OSDFState state, OpenShiftProperties properties) {
```
[the remark this answers is not in the document]
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: убрал

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 8]
```diff
@@ -3,30 +3,55 @@
...
 import io.microconfig.osdf.state.OSDFState;
+import io.microconfig.utils.StringUtils;
+import lombok.Builder;
 import lombok.RequiredArgsConstructor;
 
 import static io.microconfig.osdf.microconfig.properties.OpenShiftProperties.properties;
 import static io.microconfig.osdf.microconfig.properties.PropertyGetter.propertyGetter;
 import static io.microconfig.osdf.state.OSDFState.fromFile;
 
+@Builder
```
[the remark this answers is not in the document]
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: убрал

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java [hunk 9]
```diff
@@ -3,30 +3,55 @@
...
+        return OpenShiftProject.builder()
+                .clusterUrl(properties.clusterUrl())
+                .project(properties.project())
+                .token(state.getOpenShiftCredentials().getToken())
+                .oc(oc).build();
     }
 
     public OpenShiftProject connect() {
-        login();
+        if (StringUtils.isEmpty(token)) login(); else loginWithToken();
```
[the remark this answers is not in the document]
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: исправил

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java (hunk 1 above)
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java]: исправил

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java (hunk 2 above)
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java]: исправил

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java (hunk 3 above)
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: убрал

File: src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java (hunk 4 above)
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/openshift/OpenShiftProject.java]: Has been added

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java (hunk 5 above)
Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java:11]: Has been changed

File: src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        this.password = password;
+    }
+
+    public OpenShiftCredentials(String token) {
+        this.token = token;
+    }
+
+    public static OpenShiftCredentials of(String credentialsString) {
+        String[] split = credentialsString.split(":");
+        if (split.length == 2) return new OpenShiftCredentials(split[0], split[1]);
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java:28]: Use credentials here

Comment (slav9n4ik) [reply src/main/java/io/microconfig/osdf/state/OpenShiftCredentials.java]: done

**Pull request merged**